crates/nu-protocol/src/value/primitive.rs

```diff
@@ -284,18 +323,59 @@ pub fn format_primitive(primitive: &Primitive, field_name: Option<&String>) -> S
     }
 }
 
-/// Format a duration in seconds into a string
-pub fn format_duration(sec: i64) -> String {
-    let (minutes, seconds) = (sec / 60, sec % 60);
-    let (hours, minutes) = (minutes / 60, minutes % 60);
-    let (days, hours) = (hours / 24, hours % 24);
-
-    match (days, hours, minutes, seconds) {
-        (0, 0, 0, 1) => "1 sec".to_owned(),
-        (0, 0, 0, s) => format!("{} secs", s),
-        (0, 0, m, s) => format!("{}:{:02}", m, s),
-        (0, h, m, s) => format!("{}:{:02}:{:02}", h, m, s),
-        (d, h, m, s) => format!("{}:{:02}:{:02}:{:02}", d, h, m, s),
+/// Format a duration in nanoseconds into a string
+pub fn format_duration(duration: &BigInt) -> String {
+    // FIXME: This involves a lot of allocation, but it seems inevitable with BigInt.
+    let big_int_1000 = BigInt::from(1000);
+    let big_int_60 = BigInt::from(60);
+    let big_int_24 = BigInt::from(24);
+    // We only want the biggest subvidision to have the negative sign.
+    let (sign, duration) = if duration.is_positive() {
+        (1, duration.clone())
+    } else {
+        (-1, -duration)
+    };
+    let (microseconds, nanoseconds): (BigInt, BigInt) = duration.div_rem(&big_int_1000);
+    let (milliseconds, microseconds): (BigInt, BigInt) = microseconds.div_rem(&big_int_1000);
+    let (seconds, milliseconds): (BigInt, BigInt) = milliseconds.div_rem(&big_int_1000);
+    let (minutes, seconds): (BigInt, BigInt) = seconds.div_rem(&big_int_60);
+    let (hours, minutes): (BigInt, BigInt) = minutes.div_rem(&big_int_60);
+    let (days, hours): (BigInt, BigInt) = hours.div_rem(&big_int_24);
+
+    match (
+        days.to_i64()
+            .expect("Overflowing while formatting duration."),
+        hours.to_i64().unwrap(),
+        minutes.to_i64().unwrap(),
+        seconds.to_i64().unwrap(),
+        milliseconds.to_i64().unwrap(),
+        microseconds.to_i64().unwrap(),
+        nanoseconds.to_i64().unwrap(),
+    ) {
+        (0, 0, 0, 0, 0, 0, ns) => format!("{}", sign * ns),
+        (0, 0, 0, 0, 0, us, ns) => format!("{}:{:03}", sign * us, ns),
+        (0, 0, 0, 0, ms, us, ns) => format!("{}:{:03}:{:03}", sign * ms, us, ns),
+        (0, 0, 0, s, ms, us, ns) => format!("{}:{:03}:{:03}:{:03}", sign * s, ms, us, ns),
+        (0, 0, m, s, ms, us, ns) => format!("{}:{:02}:{:03}:{:03}:{:03}", sign * m, s, ms, us, ns),
```

Comment:
I wonder if the format should be something like:

```
{days}:{hours}:{minutes}:{sec}.{ms + us + ns}
```
So that anything smaller than a second is shown as fractions of a second.  I looked around to see if there was a standard here, but I didn't see one.